hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/qjournal/client/TestQuorumJournalManager.java

```diff
@@ -1122,41 +1115,50 @@ public void testSelectViaRpcTwoJNsError() throws Exception {
     EditLogOutputStream stm = qjm.startLogSegment(
         1, NameNodeLayoutVersion.CURRENT_LAYOUT_VERSION);
     SettableFuture<Void> slowLog = SettableFuture.create();
     Mockito.doReturn(slowLog).when(spies.get(0))
-        .sendEdits(eq(1L), eq(11L), eq(1), Mockito.any());
+        .sendEdits(eq(1L), eq(11L), eq(10), Mockito.any());
+    // Successfully write these edits to JN0 ~ JN2
     writeTxns(stm, 1, 10);
+    // Failed write these edits to JN0, but successfully write them to JN1 ~ JN2
     writeTxns(stm, 11, 10);
-    writeTxns(stm, 21, 10);
-    writeTxns(stm, 31, 10);
-    ListeningExecutorService service = MoreExecutors.listeningDecorator(
-        Executors.newSingleThreadExecutor());
-    Mockito.doAnswer(invocation -> service.submit(
-        () -> {
-          ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
-          EditLogFileOutputStream.writeHeader(
-              NameNodeLayoutVersion.CURRENT_LAYOUT_VERSION,
-              new DataOutputStream(byteStream));
-          byteStream.write(createTxnData(21, 20));
-          Thread.sleep(3000);
-          return GetJournaledEditsResponseProto.newBuilder()
-              .setTxnCount(20)
-              .setEditLog(ByteString.copyFrom(byteStream.toByteArray()))
-              .build();
-        })
-    ).when(spies.get(1)).getJournaledEdits(21,
+    // Successfully write these edits to JN1 ~ JN2
+    writeTxns(stm, 21, 20);
+
+    Semaphore semaphore = new Semaphore(0);
+
+    Mockito.doAnswer((Answer<ListenableFuture<GetJournaledEditsResponseProto>>) invocation -> {
+      semaphore.release(1);
+      @SuppressWarnings("unchecked")
+      ListenableFuture<GetJournaledEditsResponseProto> result =
+          (ListenableFuture<GetJournaledEditsResponseProto>) invocation.callRealMethod();
+      return result;
+    }).when(spies.get(0)).getJournaledEdits(21,
+        QuorumJournalManager.QJM_RPC_MAX_TXNS_DEFAULT);
+
+    Mockito.doAnswer((Answer<ListenableFuture<GetJournaledEditsResponseProto>>) invocation -> {
+      semaphore.release(1);
+      @SuppressWarnings("unchecked")
+      ListenableFuture<GetJournaledEditsResponseProto> result =
+          (ListenableFuture<GetJournaledEditsResponseProto>) invocation.callRealMethod();
+      return result;
+    }).when(spies.get(1)).getJournaledEdits(21,
         QuorumJournalManager.QJM_RPC_MAX_TXNS_DEFAULT);
 
-    GetJournaledEditsResponseProto responseProto = spies.get(2)
-        .getJournaledEdits(21, 5000).get();
-    assertEquals(20, responseProto.getTxnCount());
+    Mockito.doAnswer((Answer<ListenableFuture<GetJournaledEditsResponseProto>>) invocation -> {
+      semaphore.acquire(2);
+      @SuppressWarnings("unchecked")
+      ListenableFuture<GetJournaledEditsResponseProto> result =
+          (ListenableFuture<GetJournaledEditsResponseProto>) invocation.callRealMethod();
+      return result;
+    }).when(spies.get(2)).getJournaledEdits(21,
+        QuorumJournalManager.QJM_RPC_MAX_TXNS_DEFAULT);
```

Comment:
Nice!! Looks really good except for being pretty verbose. Can we create a helper method like:
```
private void spyGetJournaledEdits(int jnSpyIdx, long fromTxId, Runnable preHook) { ... }
```
Then here we can just have
```
    spyGetJournaledEdits(0, 21, () -> semaphore.release(1));
    spyGetJournaledEdits(1, 21, () -> semaphore.release(1));
    spyGetJournaledEdits(2, 21, () -> semaphore.acquireUninterruptibly(2));
```

Reply:
Nice suggestion, thanks.